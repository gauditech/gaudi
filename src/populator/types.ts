type PopulatorModel = {
  model: string;
  repeat: { min: number; max: number } | number;
  data: PopulatorData[];
};

type PopulatorData = {
  field: string;
  type: "constant" | "";
};
