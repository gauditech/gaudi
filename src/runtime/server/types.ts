import { NextFunction, Request, Response } from "express";

export type ServerHttpMethods = "all" | "get" | "post" | "put" | "patch" | "delete";

export type ServerRequestHandler =
  | ((request: Request, response: Response) => void | Promise<void>)
  | ((req: Request, resp: Response, next: NextFunction) => void | Promise<void>);

export type EndpointConfig = {
  path: string;
  method: ServerHttpMethods;
  handlers: ServerRequestHandler[];
};

export type ServerMiddlewareNextFn = (err?: unknown) => void;
