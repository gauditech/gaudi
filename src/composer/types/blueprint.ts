import { Model } from "./model";

export interface Blueprint {
  models: Record<string, Model>;
}
