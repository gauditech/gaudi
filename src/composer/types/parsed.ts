export interface ModelDef {
  name: string;
  fields: FieldDef[];
  references?: ReferenceDef[];
  relations?: RelationDef[];
}

export interface FieldDef {
  name: string;
  type: "string" | "integer" | "datetime";
  nullable?: boolean;
}

export interface ReferenceDef {
  name: string;
  model: string;
}

export interface RelationDef {
  name: string;
  model: string;
}
