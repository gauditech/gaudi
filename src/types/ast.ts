import { WithContext } from "@src/common/error";

export type AST = {
  models: ModelAST[];
};

export type ModelAST = WithContext<{
  name: string;
  body: ModelBodyAST[];
}>;

export type ModelBodyAST = FieldAST | ReferenceAST | RelationAST | QueryAST | ComputedAST;

export type FieldAST = WithContext<{
  kind: "field";
  name: string;
  body: FieldBodyAST[];
}>;

export type FieldBodyAST = WithContext<
  | { kind: "type"; type: string }
  | { kind: "default"; default: LiteralValue }
  | { kind: "tag"; tag: FieldTag }
>;

export type FieldTag = "nullable" | "unique";

export type ReferenceAST = WithContext<{
  kind: "reference";
  name: string;
  body: ReferenceBodyAST[];
}>;

export type ReferenceBodyAST = WithContext<
  { kind: "to"; to: string } | { kind: "tag"; tag: ReferenceTag }
>;

export type ReferenceTag = "nullable" | "unique";

export type RelationAST = WithContext<{
  kind: "relation";
  name: string;
  body: RelationBodyAST[];
}>;

export type RelationBodyAST = WithContext<
  { kind: "from"; from: string } | { kind: "through"; through: string }
>;

export type QueryAST = WithContext<{
  kind: "query";
  name: string;
  body: QueryBodyAST[];
}>;

export type QueryBodyAST = WithContext<
  | { kind: "from"; from: string[] }
  | { kind: "filter"; filter: ExpAST }
  | { kind: "orderBy"; orderings: QueryOrderAST[] }
  | { kind: "limit"; limit: number }
>;

export type QueryOrderAST = WithContext<{ field: string[]; order?: "asc" | "desc" }>;

export type ComputedAST = WithContext<{ kind: "computed"; name: string; exp: ExpAST }>;

export type ExpAST = WithContext<
  | {
      kind: "binary";
      operator: BinaryOperator;
      lhs: ExpAST;
      rhs: ExpAST;
    }
  | { kind: "paren"; exp: ExpAST }
  | { kind: "unary"; operator: UnaryOperator; exp: ExpAST }
  | { kind: "identifier"; identifier: string[] }
  | { kind: "literal"; literal: LiteralValue }
>;

export type Entrypoint = WithContext<{
  name: string;
  body: EntrypointBody[];
}>;

export type EntrypointBody = WithContext<
  | { kind: "target"; identifier: string }
  | { kind: "identify"; identifier: string }
  | { kind: "alias"; identifier: string }
  | { kind: "response"; select: string[] }
  | { kind: "endpoint"; endpoint: Endpoint }
  | { kind: "entrypoint"; entrypoint: Entrypoint }
>;

export type Endpoint = WithContext<{
  type: EndpointType;
  body: EndpointBody;
}>;

export type EndpointType = "list" | "get" | "create" | "update" | "delete";

export type EndpointBody = WithContext<{ kind: "action" }>;

export type LiteralValue = null | boolean | number | string;

export type BinaryOperator =
  | "or"
  | "and"
  | "is not"
  | "is"
  | "not in"
  | "in"
  | "<"
  | "<="
  | ">"
  | ">=";

export type UnaryOperator = "not";
