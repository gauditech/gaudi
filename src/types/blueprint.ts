import { Field, Model } from "./model";

export interface Blueprint {
  models: Record<string, Model>;
  fields: Record<string, Field>;
}
