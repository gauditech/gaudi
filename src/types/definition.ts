export type Definition = {
  models: ModelDef[];
};

export type ModelDef = {
  refKey: string;
  name: string;
  dbname: string;
  fields: FieldDef[];
  references: ReferenceDef[];
  relations: RelationDef[];
  queries: QueryDef[];
};

export type FieldDef = {
  refKey: string;
  modelRefKey: string;
  name: string;
  dbname: string;
  type: "integer" | "text" | "boolean";
  dbtype: "serial" | "integer" | "text" | "boolean";
  primary: boolean;
  unique: boolean;
  nullable: boolean;
};

export type ReferenceDef = {
  refKey: string;
  modelRefKey: string;
  name: string;
  fieldRefKey: string;
  toModelRefKey: string;
  toModelFieldRefKey: string;
  nullable: boolean;
  unique: boolean;
};

export type RelationDef = {
  refKey: string;
  modelRefKey: string;
  name: string;
  fromModel: string;
  fromModelRefKey: string;
  through: string;
  throughRefKey: string;
  nullable: boolean;
  unique: boolean;
};

type QueryDefPathSelect = {
  name: string;
  retType: string;
  nullable: boolean;
  refKey: string;
};

export type QueryDefPath = {
  refKey: string;
  name: string;
  retType: string;
  retCardinality: "one" | "many"; // should be retCard,...
  nullable: boolean;
  joinType: "inner" | "left";
  alias: string;
  bpAlias: string | null;
  joinPaths: QueryDefPath[];
  select: QueryDefPathSelect[];
};

type QueryDefFilter = {
  type: string;
  lhs: string | number | boolean;
  rhs: string | number | boolean;
};

export type QueryDef = {
  refKey: string;
  name: string;
  retType: string;
  retCardinality: "one" | "many";
  nullable: boolean;
  // unique: boolean;
  joinPaths: QueryDefPath[];
  filters: QueryDefFilter[];
};
