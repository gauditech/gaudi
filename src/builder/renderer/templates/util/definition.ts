import { Definition, PathFragment } from "@src/types/definition";

/*
 * Template helper functions for Gaudi definition
 */

export function modelDbName(modelRefKey: string, definition: Definition) {
  return (definition.models || []).find((model) => model.refKey === modelRefKey)?.dbname;
}

export function fieldDbName(modelRefKey: string, fieldRefKey: string, definition: Definition) {
  return (definition.models || [])
    .find((model) => model.refKey === modelRefKey)
    ?.fields.find((field) => field.refKey === fieldRefKey)?.dbname;
}

export function typeToDbType(type: string) {
  if (type === "serial") {
    return "Int";
  } else if (type === "integer") {
    return "Int";
  } else if (type === "text") {
    return "String";
  } else if (type === "boolean") {
    return "Boolean";
  }

  throw "UNHANDLED_GAUDI_TYPE";
}

export function buildEndpointPath(parts: PathFragment[]): string {
  return parts.reduce((accum, part) => {
    const pathPart =
      part.type === "numeric" || part.type === "text" ? `:${part.varname}` : part.value;
    return `${accum}/${pathPart}`;
  }, "");
}
