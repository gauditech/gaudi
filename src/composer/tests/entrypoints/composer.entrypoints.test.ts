import { compileToOldSpec, compose } from "@src/index";
import { CreateEndpointDef, CustomManyEndpointDef, ExecuteHookAction } from "@src/types/definition";

describe("entrypoint", () => {
  it("composes basic example", () => {
    // Orgs assumes default response
    // Orgs.Repositories assumes default identifyWith; nested org select assuming all fields since not given
    const bp = `
    model Org {
      field slug { type string, unique }
      field name { type string }
      relation repos { from Repo, through org }
    }
    model Repo {
      reference org { to Org }
      field title { type string }
    }

    entrypoint Orgs {
      target Org
      identify with slug

      list endpoint {}
      get endpoint {}

      entrypoint Repositories {
        target repos as repo
        response { id, org }

        list endpoint {}
        get endpoint {}
        create endpoint {}

        custom endpoint {
          cardinality one
          method PATCH
          path "somePath"

          action {
            update {}
          }
        }
      }
    }
    `;
    const def = compose(compileToOldSpec(bp));
    expect(def.entrypoints).toMatchSnapshot();
  });
  it("adds validators into fieldsets", () => {
    const bp = `
    model Org {
      field name { type string, validate { min 4, max 100 } }
    }

    entrypoint Orgs {
      target Org
      create endpoint {}
    }
    `;
    const def = compose(compileToOldSpec(bp));
    const endpoint = def.entrypoints[0].endpoints[0] as CreateEndpointDef;
    expect(endpoint.fieldset).toMatchSnapshot();
  });

  it("action should send response", () => {
    const bp = `
    runtime MyRuntime {
      source path "some/source/path"
    }

    model Org {}

    entrypoint Orgs {
      target Org

      // endpoint W/ responding action
      custom endpoint {
        path "somePath1"
        method POST
        cardinality many

        action {
          execute {
            responds
            hook {
              runtime MyRuntime
              source testFn from "t/h/p"

            }
          }
        }
      }
    }
    `;
    const def = compose(compileToOldSpec(bp));

    const endpoint = def.entrypoints[0].endpoints[0] as CustomManyEndpointDef;
    const action = endpoint.actions[0] as ExecuteHookAction;

    expect(action.responds).toBe(true);
    expect(endpoint.responds).toBe(false);
  });

  it("endpoint should send response", () => {
    const bp = `
    runtime MyRuntime {
      source path "some/source/path"
    }

    model Org {}

    entrypoint Orgs {
      target Org

      // endpoint W/O responding action
      custom endpoint {
        path "somePath1"
        method POST
        cardinality many

        action {
          execute {
            hook {
              runtime MyRuntime
              source testFn from "t/h/p"
            }
          }
        }
      }
    }
    `;
    const def = compose(compileToOldSpec(bp));

    const endpoint = def.entrypoints[0].endpoints[0] as CustomManyEndpointDef;
    const action = endpoint.actions[0] as ExecuteHookAction;

    expect(action.responds).toBe(false);
    expect(endpoint.responds).toBe(true);
  });

  it("fail for multiple actions that want to respond", () => {
    const bp = `
    runtime MyRuntime {
      source path "some/source/path"
    }

    model Org {}

    entrypoint Orgs {
      target Org

      custom endpoint {
        path "somePath1"
        method POST
        cardinality many

        action {
          execute {
            responds
            hook {
              runtime MyRuntime
              source testFn from "t/h/p"
            }
          }
          execute {
            responds
            hook {
              runtime MyRuntime
              source testFn from "t/h/p"
            }
          }
        }
      }
    }
    `;

    expect(() => compose(compileToOldSpec(bp))).toThrowErrorMatchingInlineSnapshot(
      `"At most one action in entrypoint can have "responds" attribute"`
    );
  });

  it("fails if responds action is used in implicit endpoints", () => {
    const bp = `
    runtime MyRuntime {
      source path "some/source/path"
    }

    model Org {}

    entrypoint Orgs {
      target Org

      create endpoint {
        action {
          execute {
            responds
            hook {
              runtime MyRuntime
              source testFn from "t/h/p"
            }
          }
        }
      }
    }
    `;

    expect(() => compose(compileToOldSpec(bp))).toThrowErrorMatchingInlineSnapshot(
      `"Actions with "responds" keyword are allowed only in "custom" endpoints, not in "create""`
    );
  });

  it("fails if responds action is used in implicit actions", () => {
    const bp = `
    model Org {}

    entrypoint Orgs {
      target Org

      custom endpoint {
        path "somePath"
        method POST
        cardinality many

        action {
          create {
            responds
          }
        }
      }
    }
    `;

    expect(() => compileToOldSpec(bp)).toThrowErrorMatchingInlineSnapshot(`
      "Unknown source position!
      responds is not a valid model action"
    `);
  });
  it("collects dependencies", () => {
    const bp = `
    model Org {
      field name { type string }
      field desc { type string }

      relation repos { from Repo, through org }

      query repoCount { from repos, count }
      computed coef { 2 }
    }
    model Repo {
      reference org { to Org }

      field name { type string }

      relation issues { from Issue, through repo}
    }
    model Issue {
      reference repo { to Repo }

      field source { type string }
      field orgDesc { type string }
      field orgCoef { type integer }
    }

    entrypoint O {
      target Org as org
      entrypoint R {
        target repos as repo
        create endpoint {
          action {
            create {}
            create repo.issues as i {
              set source concat(org.name, repo.name)
              set orgDesc org.desc
              set orgCoef org.repoCount * org.coef
            }
          }
        }
      }
    }
    `;
    const def = compose(compileToOldSpec(bp));
    const endpoint = def.entrypoints[0].entrypoints[0].endpoints[0] as CreateEndpointDef;
    const orgSelect = endpoint.parentContext[0].select.map((s) => s.alias);
    const repoSelect = endpoint.target.select.map((s) => s.alias);
    const actionSelects = endpoint.actions.map((a) =>
      ("select" in a ? a.select : null)?.map((a) => a && a.alias)
    );
    expect({ orgSelect, repoSelect, actionSelects }).toMatchSnapshot();
  });
});
