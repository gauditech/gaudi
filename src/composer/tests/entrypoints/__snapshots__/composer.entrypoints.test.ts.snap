// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entrypoint adds validators into fieldsets 1`] = `
{
  "kind": "record",
  "nullable": false,
  "record": {
    "name": {
      "kind": "field",
      "nullable": false,
      "required": true,
      "type": "text",
      "validators": [
        {
          "args": [
            {
              "type": "integer",
              "value": 4,
            },
          ],
          "inputType": "text",
          "name": "minLength",
        },
        {
          "args": [
            {
              "type": "integer",
              "value": 100,
            },
          ],
          "inputType": "text",
          "name": "maxLength",
        },
      ],
    },
  },
}
`;

exports[`entrypoint composes basic example 1`] = `
[
  {
    "endpoints": [
      {
        "kind": "list",
        "parentContext": [],
        "response": [
          {
            "alias": "id",
            "kind": "field",
            "name": "id",
            "namePath": [
              "Org",
              "id",
            ],
            "refKey": "Org.id",
          },
          {
            "alias": "slug",
            "kind": "field",
            "name": "slug",
            "namePath": [
              "Org",
              "slug",
            ],
            "refKey": "Org.slug",
          },
          {
            "alias": "name",
            "kind": "field",
            "name": "name",
            "namePath": [
              "Org",
              "name",
            ],
            "refKey": "Org.name",
          },
        ],
        "target": {
          "alias": "$target_0",
          "kind": "model",
          "name": "Org",
          "namePath": [
            "Org",
          ],
          "refKey": "Org",
          "retType": "Org",
          "select": [],
        },
      },
      {
        "kind": "get",
        "parentContext": [],
        "response": [
          {
            "alias": "id",
            "kind": "field",
            "name": "id",
            "namePath": [
              "Org",
              "id",
            ],
            "refKey": "Org.id",
          },
          {
            "alias": "slug",
            "kind": "field",
            "name": "slug",
            "namePath": [
              "Org",
              "slug",
            ],
            "refKey": "Org.slug",
          },
          {
            "alias": "name",
            "kind": "field",
            "name": "name",
            "namePath": [
              "Org",
              "name",
            ],
            "refKey": "Org.name",
          },
        ],
        "target": {
          "alias": "$target_0",
          "identifyWith": {
            "name": "slug",
            "paramName": "org_slug",
            "refKey": "Org.slug",
            "type": "text",
          },
          "kind": "model",
          "name": "Org",
          "namePath": [
            "Org",
          ],
          "refKey": "Org",
          "retType": "Org",
          "select": [],
        },
      },
    ],
    "entrypoints": [
      {
        "endpoints": [
          {
            "kind": "list",
            "parentContext": [
              {
                "alias": "$target_0",
                "identifyWith": {
                  "name": "slug",
                  "paramName": "org_slug",
                  "refKey": "Org.slug",
                  "type": "text",
                },
                "kind": "model",
                "name": "Org",
                "namePath": [
                  "Org",
                ],
                "refKey": "Org",
                "retType": "Org",
                "select": [],
              },
            ],
            "response": [
              {
                "alias": "id",
                "kind": "field",
                "name": "id",
                "namePath": [
                  "Org",
                  "repos",
                  "id",
                ],
                "refKey": "Repo.id",
              },
              {
                "alias": "org",
                "kind": "reference",
                "name": "org",
                "namePath": [
                  "Org",
                  "repos",
                  "org",
                ],
                "select": [
                  {
                    "alias": "id",
                    "kind": "field",
                    "name": "id",
                    "namePath": [
                      "Org",
                      "repos",
                      "org",
                      "id",
                    ],
                    "refKey": "Org.id",
                  },
                  {
                    "alias": "slug",
                    "kind": "field",
                    "name": "slug",
                    "namePath": [
                      "Org",
                      "repos",
                      "org",
                      "slug",
                    ],
                    "refKey": "Org.slug",
                  },
                  {
                    "alias": "name",
                    "kind": "field",
                    "name": "name",
                    "namePath": [
                      "Org",
                      "repos",
                      "org",
                      "name",
                    ],
                    "refKey": "Org.name",
                  },
                ],
              },
            ],
            "target": {
              "alias": "repo",
              "kind": "relation",
              "name": "repos",
              "namePath": [
                "Org",
                "repos",
              ],
              "refKey": "Org.repos",
              "retType": "Repo",
              "select": [],
            },
          },
          {
            "kind": "get",
            "parentContext": [
              {
                "alias": "$target_0",
                "identifyWith": {
                  "name": "slug",
                  "paramName": "org_slug",
                  "refKey": "Org.slug",
                  "type": "text",
                },
                "kind": "model",
                "name": "Org",
                "namePath": [
                  "Org",
                ],
                "refKey": "Org",
                "retType": "Org",
                "select": [],
              },
            ],
            "response": [
              {
                "alias": "id",
                "kind": "field",
                "name": "id",
                "namePath": [
                  "Org",
                  "repos",
                  "id",
                ],
                "refKey": "Repo.id",
              },
              {
                "alias": "org",
                "kind": "reference",
                "name": "org",
                "namePath": [
                  "Org",
                  "repos",
                  "org",
                ],
                "select": [
                  {
                    "alias": "id",
                    "kind": "field",
                    "name": "id",
                    "namePath": [
                      "Org",
                      "repos",
                      "org",
                      "id",
                    ],
                    "refKey": "Org.id",
                  },
                  {
                    "alias": "slug",
                    "kind": "field",
                    "name": "slug",
                    "namePath": [
                      "Org",
                      "repos",
                      "org",
                      "slug",
                    ],
                    "refKey": "Org.slug",
                  },
                  {
                    "alias": "name",
                    "kind": "field",
                    "name": "name",
                    "namePath": [
                      "Org",
                      "repos",
                      "org",
                      "name",
                    ],
                    "refKey": "Org.name",
                  },
                ],
              },
            ],
            "target": {
              "alias": "repo",
              "identifyWith": {
                "name": "id",
                "paramName": "repo_id",
                "refKey": "Repo.id",
                "type": "integer",
              },
              "kind": "relation",
              "name": "repos",
              "namePath": [
                "Org",
                "repos",
              ],
              "refKey": "Org.repos",
              "retType": "Repo",
              "select": [],
            },
          },
          {
            "actions": [
              {
                "alias": "repo",
                "changeset": [
                  {
                    "name": "title",
                    "setter": {
                      "fieldsetAccess": [
                        "title",
                      ],
                      "kind": "fieldset-input",
                      "required": true,
                      "type": "text",
                    },
                  },
                  {
                    "name": "org_id",
                    "setter": {
                      "kind": "reference-value",
                      "target": {
                        "access": [
                          "id",
                        ],
                        "alias": "$target_0",
                      },
                    },
                  },
                ],
                "kind": "create-one",
                "model": "Repo",
                "select": [],
                "targetPath": [
                  "repo",
                ],
              },
            ],
            "fieldset": {
              "kind": "record",
              "nullable": false,
              "record": {
                "title": {
                  "kind": "field",
                  "nullable": false,
                  "required": true,
                  "type": "text",
                  "validators": [],
                },
              },
            },
            "kind": "create",
            "parentContext": [
              {
                "alias": "$target_0",
                "identifyWith": {
                  "name": "slug",
                  "paramName": "org_slug",
                  "refKey": "Org.slug",
                  "type": "text",
                },
                "kind": "model",
                "name": "Org",
                "namePath": [
                  "Org",
                ],
                "refKey": "Org",
                "retType": "Org",
                "select": [
                  {
                    "alias": "id",
                    "kind": "field",
                    "name": "id",
                    "namePath": [
                      "Org",
                      "id",
                    ],
                    "refKey": "Org.id",
                  },
                ],
              },
            ],
            "response": [
              {
                "alias": "id",
                "kind": "field",
                "name": "id",
                "namePath": [
                  "Org",
                  "repos",
                  "id",
                ],
                "refKey": "Repo.id",
              },
              {
                "alias": "org",
                "kind": "reference",
                "name": "org",
                "namePath": [
                  "Org",
                  "repos",
                  "org",
                ],
                "select": [
                  {
                    "alias": "id",
                    "kind": "field",
                    "name": "id",
                    "namePath": [
                      "Org",
                      "repos",
                      "org",
                      "id",
                    ],
                    "refKey": "Org.id",
                  },
                  {
                    "alias": "slug",
                    "kind": "field",
                    "name": "slug",
                    "namePath": [
                      "Org",
                      "repos",
                      "org",
                      "slug",
                    ],
                    "refKey": "Org.slug",
                  },
                  {
                    "alias": "name",
                    "kind": "field",
                    "name": "name",
                    "namePath": [
                      "Org",
                      "repos",
                      "org",
                      "name",
                    ],
                    "refKey": "Org.name",
                  },
                ],
              },
            ],
            "target": {
              "alias": "repo",
              "kind": "relation",
              "name": "repos",
              "namePath": [
                "Org",
                "repos",
              ],
              "refKey": "Org.repos",
              "retType": "Repo",
              "select": [],
            },
          },
        ],
        "entrypoints": [],
        "name": "Repositories",
        "target": {
          "alias": "repo",
          "identifyWith": {
            "name": "id",
            "paramName": "repo_id",
            "refKey": "Repo.id",
            "type": "integer",
          },
          "kind": "relation",
          "name": "repos",
          "namePath": [
            "Org",
            "repos",
          ],
          "refKey": "Org.repos",
          "retType": "Repo",
        },
      },
    ],
    "name": "Orgs",
    "target": {
      "alias": "$target_0",
      "identifyWith": {
        "name": "slug",
        "paramName": "org_slug",
        "refKey": "Org.slug",
        "type": "text",
      },
      "kind": "model",
      "name": "Org",
      "namePath": [
        "Org",
      ],
      "refKey": "Org",
      "retType": "Org",
    },
  },
]
`;
